README
==============================================================
wdb-stationLoad --- loading weather station meta-data into WDB
==============================================================

Norwegian Meteorological Institute (met.no)
Box 43 Blindern
0313 OSLO
NORWAY

email: aleksandarb@met.no

WDB is a database system for storage and retrieval of weather and 
water observations and forecasts. It handles both gridded data
(raster) and point data.

wdb-stationLoad:
    Loads weather stations meta-data (location, validity time ...) into the WDB database. 
    The original data is read from STINFOSYS database. 
    The wdb-stationLoad operates straightforward: reads the records from STINFOSYS and inserts them into WDB. 
    This way the stations history should be preserved in WDB as well. 
    The WDB is expected to handle the meta-data properly and consitently. 
    The STINFOSYS is expected to provide correct data.


To compile and run app make sure you have wdb libraries.
Please check that WDB version supports wci functions like:
    wci.addplacepoint
    wci.addorupdateplacepoint

The configure line might look like:
    ./configure wdbload_CFLAGS=-I<local-dir>/include LDFLAGS=-L<local-dir>/lib wdbload_LIBS=-lwdbLoad

STINFOSYS (see: http://stinfosys/) will be used as authoritative source for weather stations meta-data.

When building debian packages it is importan to have proper setup.
This mean that all wdb dependenciea are in place. If wdb has been
installed then include and library paths will point to it.


USAGE
       wdb-stationLoad can be called with all options present on the command line (check --help or man pages).
       It is more convenient to use main configuration file to set the values:

       wdb-stationLoad --config /etc/stationload.conf

       Finer logging can be turned on to see the data that is to be inserted into WDB:

       wdb-stationLoad --config /etc/stationload.conf --loglevel 1

       Main config file looks like:

           cns-namespace=88000
           wmo-namespace=88001

           # WDB database connection parameters
           database =     # database name
           host =         # host name
           port =         # port
           user =         # user name

           # STINFOSYS database connection parameters
           stdatabase =   # database name
           sthost =       # host name
           stuser =       # user name
           stport =       # port
           stpass =       # password


ChangeLog
WDB GNU ChangeLog

This file is not maintained. To get a GNU-style ChangeLog for WDB, use
svn log on the WDB system's git repository.


src/STLoaderConfiguration.cpp
/*
    wdb - weather and water data storage

    Copyright (C) 2012 met.no

    Contact information:
    Norwegian Meteorological Institute
    Box 43 Blindern
    0313 OSLO
    NORWAY
    E-mail: wdb@met.no

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
    MA  02110-1301, USA
*/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "STLoaderConfiguration.h"

// boost
#include <boost/program_options.hpp>

using namespace std;
using namespace boost::program_options;

// Support Functions
namespace
{

options_description getOutput(wdb::load::STLoaderConfiguration::OutputOptions & out)
{
    options_description output( "Output" );
    output.add_options()
       ( "dry-run", bool_switch( & out.dry_run ), "Just show the SQL commands that would be sent to WDB database" )
    ;

    return output;
}

options_description getLoading(wdb::load::STLoaderConfiguration::LoadingOptions & out, const std::string & defaultDataProvider )
{
    out.defaultDataProvider = defaultDataProvider;
    options_description loading("Station Loading");
    loading.add_options()
        ( "stdatabase", value( & out.stdatabase ), "Specify stations database name" )
        ( "sthost", value( & out.sthost ), "Specify stations database host" )
        ( "stuser", value( & out.stuser ), "Specify stations database user" )
        ( "stpass", value( & out.stpass ), "Specify stations database password" )
        ( "stport", value( & out.stport )->default_value( 5432 ), "Specify stations database port (default 5432)" )
        ( "cns-namespace", value( & out.cnsNamespace )->default_value( 88000 ), "Specify stationid namespace (default 88000)" )
        ( "wmo-namespace", value( & out.wmoNamespace )->default_value( 88001 ), "Specify WMO namespace (default 88001)" )
        ( "after", value( & out.stupdatedafter ), "Specify date of last update for stations (=> after)" )
    ;

    return loading;
}

} // namespace support functions

namespace wdb { namespace load {
STLoaderConfiguration::STLoaderConfiguration(const std::string & defaultDataProvider)
        : WdbConfiguration(""), defaultDataProvider_(defaultDataProvider)
{
        configOptions().add( getOutput( output_ ) );
        configOptions().add( getLoading( loading_, defaultDataProvider_ ) );

        shownOptions().add( getOutput( output_ ) );
        shownOptions().add( getLoading( loading_, defaultDataProvider_  ) );
}

STLoaderConfiguration::~STLoaderConfiguration() { }

string STLoaderConfiguration::LoadingOptions::pqDatabaseConnection() const
{
        ostringstream ret;
        ret << "dbname=" << stdatabase.c_str() << " ";
        if(sthost != "")
            ret << "host=" << sthost.c_str() << " ";
        ret << "port=" << stport << " ";
        if(stuser != "")
            ret << "user=" << stuser.c_str() << " ";
        if(stpass != "")
            ret << "password=" << stpass.c_str() << " ";
        return ret.str();
}

string STLoaderConfiguration::LoadingOptions::psqlDatabaseConnection() const
{
        ostringstream ret;
        ret << "-d" << stdatabase;
        if (!sthost.empty())
            ret << " -h" << sthost;
        ret << " -p" << stport;
        ret << " -U" << stuser;
        return ret.str();
}
} }
